Mathlib/Util/PiNotation.lean

```diff
@@ -42,12 +42,33 @@ parse it by simply using the pre-existing forall parser. -/
   | .node info _ args => return .node info ``Lean.Parser.Term.forall args
   | _ => Lean.Macro.throwUnsupported
 
-/-- Override the Lean 4 pi notation delaborator with one that uses `Π`.
+/-- Override the Lean 4 pi notation delaborator with one that prints cute binders
+such as `∀ ε > 0`. -/
+@[delab forallE]
+def delabPi : Delab := whenPPOption Lean.getPPNotation do
+  let stx ← delabForall
+  match stx with
+  | `(∀ ($i:ident : $_), $j:ident ∈ $s → $body) =>
+    if i == j then `(∀ $i:ident ∈ $s, $body) else pure stx
+  | `(∀ ($x:ident : $_), $y:ident > $z → $body) =>
+    if x == y then `(∀ $x:ident > $z, $body) else pure stx
+  | `(∀ ($x:ident : $_), $y:ident < $z → $body) =>
+    if x == y then `(∀ $x:ident < $z, $body) else pure stx
+  | `(∀ ($x:ident : $_), $y:ident ≥ $z → $body) =>
+    if x == y then `(∀ $x:ident ≥ $z, $body) else pure stx
+  | `(∀ ($x:ident : $_), $y:ident ≤ $z → $body) =>
+    if x == y then `(∀ $x:ident ≤ $z, $body) else pure stx
+  | `(Π ($i:ident : $_), $j:ident ∈ $s → $body) =>
+    if i == j then `(Π $i:ident ∈ $s, $body) else pure stx
+  | _ => pure stx
+
+/-- Override the Lean 4 pi notation delaborator with one that uses `Π` and prints
+cute binders such as `∀ ε > 0`.
 Note that this takes advantage of the fact that `(x : α) → p x` notation is
 never used for propositions, so we can match on this result and rewrite it. -/
 @[scoped delab forallE]
-def delabPi : Delab := whenPPOption Lean.getPPNotation do
-  let stx ← delabForall
+def delabPi' : Delab := whenPPOption Lean.getPPNotation do
```

Comment:
Why do we have two pi delaborators here?

Reply:
Because some people actually like `(x : α) → p x` instead of `Π (x : α), p x` so the nice one is scoped.